### ISSUE TITLE
Integer overflow when parsing atom count in molecular formula

### ISSUE THREAD

**Issue by oatkachenko** [cdk/cdk#733]
### If molecular formula has huge amount of atoms then parser calculates it improperly.

Molecular formula: _H4294967296_
**AR**: 0 atoms
**ER**: 1 atom

Molecular formula: _H4294967298_
**AR**: 2 atoms
**ER**: 1 atom

**Comment by johnmay** [comment]
That is more atoms than the largest known molecule,  I think it's fine.

C169719H270466N45688O52238S911
https://en.wikipedia.org/wiki/Titin

Even if we use a long instead of an integer you are likely to Run out of RAM as the parse actually constructs the atoms.

**Comment by oatkachenko** [comment]
I pushed commit before I saw your answer :) Anyway thank you for explanation.

**Comment by johnmay** [comment]
Almost there, but why not just return out of the loop:

```
while (isDigit(c = next())) {
   res = (10 * res) + (c - '0');
   if (res < 0) return -1;
}
```

You do the check after the loop but you can overflow back to 0

**Comment by oatkachenko** [comment]
In this case, if overflow will happen in the middle of a number - iterator will stay there and we can't get the other part of the string.

**Comment by johnmay** [comment]
If it overflows it should be an error non? "Unexpected input"

**Comment by johnmay** [comment]
So the rest of the string doesn't matter, if someone wants to handle formulas with more than 2 billion hydrogens I think they can write their own parser for that :-)

**Comment by oatkachenko** [comment]
> If it overflows it should be an error non? "Unexpected input"

I think it should

```
while (isDigit(c = next())) {
   res = (10 * res) + (c - '0');
   if (res < 0) return -1;
}
```
btw, it is possible to overflow from positive to positive e.g. **4294967298** will be _429496729_ and then _2_

> So the rest of the string doesn't matter, if someone wants to handle formulas with more than 2 billion hydrogens I think they can write their own parser for that :-)

Absolutely agree:)

**Comment by oatkachenko** [comment]
What are the advantages to return 1 if amount of atom is negative value?

**MolecularFormulaManipulator#parseIsotope**
```
count = iter.nextUInt();
        if (count < 0)
            count = 1;
```

**Comment by johnmay** [comment]
I'd have to remember what the old code did but yes count < 0 should probably just be an error

**Comment by johnmay** [comment]
Fixed via #808
